File: src/norefund/core/costing.py
"""Pure cost and context-window calculations."""

from __future__ import annotations

import math

from norefund.core.models_registry import ModelInfo

# Tokens reserved for model output within the context window when chunking
_RESERVED_OUTPUT = 1_024


def context_usage_pct(token_count: int, context_window: int) -> float | None:
    """Return percentage of context window used (rounded to 2 dp).

    Returns None (not 0) when context_window is zero or negative so the
    caller/GUI can display '—' rather than a misleading 0 %.
    """
    if context_window <= 0:
        return None
    return round(token_count / context_window * 100, 2)


def fits_in_context(token_count: int, context_window: int) -> bool:
    if context_window <= 0:
        return False
    return token_count <= context_window


def min_chunks(token_count: int, context_window: int) -> int:
    """Minimum API calls needed to process the full document."""
    if context_window <= 0:
        return 0
    usable = max(context_window - _RESERVED_OUTPUT, 1)
    return math.ceil(token_count / usable)


def _long_context_active(prompt_token_count: int, model: ModelInfo) -> bool:
    """Whether prompt_token_count crosses this model's long-context tier.

    The tier is decided by the prompt (input) size, per every provider that
    offers one -- not by the completion size, even when pricing output.
    """
    return (
        model.long_context_threshold is not None
        and prompt_token_count > model.long_context_threshold
    )


def input_cost(token_count: int, model: ModelInfo) -> float:
    """USD cost for processing token_count input/prompt tokens."""
    rate = model.input_price_per_million
    if (
        _long_context_active(token_count, model)
        and model.long_context_input_price_per_million is not None
    ):
        rate = model.long_context_input_price_per_million
    return (token_count / 1_000_000) * rate


def output_cost(
    token_count: int, model: ModelInfo, *, prompt_token_count: int | None = None
) -> float:
    """USD cost for token_count output tokens.

    prompt_token_count is the input/prompt size that decides which price
    tier applies. Defaults to token_count itself when omitted, which is
    only correct for flat-priced models (long_context_threshold is None) --
    callers pricing a tiered model must pass the real prompt size.
    """
    prompt_size = token_count if prompt_token_count is None else prompt_token_count
    rate = model.output_price_per_million
    if (
        _long_context_active(prompt_size, model)
        and model.long_context_output_price_per_million is not None
    ):
        rate = model.long_context_output_price_per_million
    return (token_count / 1_000_000) * rate


def total_cost(input_tokens: int, output_tokens: int, model: ModelInfo) -> float:
    return input_cost(input_tokens, model) + output_cost(
        output_tokens, model, prompt_token_count=input_tokens
    )

File: frontend/src/lib/costing.ts
// Pure cost/context math, ported from core/costing.py. Recompute must be
// synchronous (no bridge round-trip) since the Calculator recomputes on
// every keystroke -- see GUI_REBUILD/05-CALCULATOR.md. Any edge-case
// behaviour here (context_window <= 0, etc.) must match the Python exactly;
// tests in costing.test.ts assert parity with tests/test_costing.py.
import type { ExchangeRates, ModelInfo } from "./types";

// Tokens reserved for model output within the context window when chunking.
const RESERVED_OUTPUT = 1_024;

/** Percentage of the context window used, rounded to 2dp. `null` (not 0)
 *  when the window is zero or negative, so callers show "—" rather than a
 *  misleading 0%. */
export function contextUsagePct(tokens: number, window: number): number | null {
  if (window <= 0) return null;
  return Math.round((tokens / window) * 100 * 100) / 100;
}

export function fitsInContext(tokens: number, window: number): boolean {
  if (window <= 0) return false;
  return tokens <= window;
}

/** Minimum API calls needed to process the full document. */
export function minChunks(tokens: number, window: number): number {
  if (window <= 0) return 0;
  const usable = Math.max(window - RESERVED_OUTPUT, 1);
  return Math.ceil(tokens / usable);
}

type TieredPricingFields = Pick<
  ModelInfo,
  | "long_context_threshold"
  | "long_context_input_price_per_million"
  | "long_context_output_price_per_million"
>;

/** Whether promptTokens crosses this model's long-context tier -- decided by
 *  the prompt (input) size, per every provider that offers one, not by the
 *  completion size even when pricing output. */
function longContextActive(promptTokens: number, model: TieredPricingFields): boolean {
  return (
    model.long_context_threshold !== null && promptTokens > model.long_context_threshold
  );
}

export function inputCost(
  tokens: number,
  model: Pick<ModelInfo, "input_price_per_million"> & TieredPricingFields,
): number {
  const rate =
    longContextActive(tokens, model) && model.long_context_input_price_per_million !== null
      ? model.long_context_input_price_per_million
      : model.input_price_per_million;
  return (tokens / 1_000_000) * rate;
}

/** promptTokens is the input size that decides which price tier applies.
 *  Defaults to tokens itself when omitted, which is only correct for
 *  flat-priced models (long_context_threshold null) -- callers pricing a
 *  tiered model must pass the real prompt size. */
export function outputCost(
  tokens: number,
  model: Pick<ModelInfo, "output_price_per_million"> & TieredPricingFields,
  promptTokens?: number,
): number {
  const promptSize = promptTokens ?? tokens;
  const rate =
    longContextActive(promptSize, model) &&
    model.long_context_output_price_per_million !== null
      ? model.long_context_output_price_per_million
      : model.output_price_per_million;
  return (tokens / 1_000_000) * rate;
}

export function totalCost(
  inputTokens: number,
  outputTokens: number,
  model: Pick<ModelInfo, "input_price_per_million" | "output_price_per_million"> &
    TieredPricingFields,
): number {
  return inputCost(inputTokens, model) + outputCost(outputTokens, model, inputTokens);
}

/** Convert a USD amount into toCurrency using the given rates -- mirrors
 *  core/currency.py's convert() exactly, including falling through
 *  unconverted (rate 1.0) for a currency the cache doesn't have a rate for,
 *  rather than throwing. */
export function convertCurrency(
  amountUsd: number,
  toCurrency: string,
  rates: ExchangeRates,
): number {
  const rate = rates.rates[toCurrency] ?? 1.0;
  return amountUsd * rate;
}
